### PR TITLE
Nb perf tests

### PR DESCRIPTION
**Pull request by kseager**
+enables stand-alone build for perf tests
+adds non-blocking tests
+fixes shmem_free bug that showed up in Cray SHMEM runs

@jdinan or @sayantansur for review

### CODE REVIEW

File: test/performance/shmem_perf_suite/shmem_bibw_nb_get_perf.c [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+    shmem_barrier_all();
+
+    if (metric_info->my_node % 2 == 0) {
+        for (i = 0; i < metric_info->trials + metric_info->warmup; i++) {
+            if(i == metric_info->warmup)
+                start = shmemx_wtime();
+
+            for(j = 0; j < metric_info->window_size; j++)
+                shmem_getmem_nbi(metric_info->dest, metric_info->src, len, dest);
```
Comment (sayantansur) [inline test/performance/shmem_perf_suite/shmem_bibw_nb_get_perf.c]: Should we put a synchronization point after every window operations? My concern is that one PE can 'race' ahead of the other and this can lead to the test reporting a bibw that can be greater than the network max (since transactions don't necessarily hit the wire at the same time).

File: test/performance/shmem_perf_suite/shmem_bibw_nb_put_perf.c [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+    shmem_barrier_all();
+
+    if (metric_info->my_node % 2 == 0) {
+        for (i = 0; i < metric_info->trials + metric_info->warmup; i++) {
+            if(i == metric_info->warmup)
+                start = shmemx_wtime();
+
+            for(j = 0; j < metric_info->window_size; j++)
+                shmem_putmem_nbi(metric_info->dest, metric_info->src, len, dest);
```
Comment (sayantansur) [inline test/performance/shmem_perf_suite/shmem_bibw_nb_put_perf.c]: same sync comment applies here.

File: test/performance/shmem_perf_suite/common.h [hunk 3]
```diff
@@ -14,6 +14,21 @@
...
+{
+    double wtime;
+    struct timeval tv;
+
+    gettimeofday(&tv, NULL);
+    wtime = tv.tv_sec;
+    wtime += (double)tv.tv_usec / 1000000.0;
+    return wtime;
+}
+#endif
```
Comment (sayantansur) [inline test/performance/shmem_perf_suite/common.h]: I don't know what shmemx_wtime() is supposed to return. Seconds? This will lose the timer resolution. Should we be reporting ns instead?

Comment (jdinan) [inline test/performance/shmem_perf_suite/common.h]: This returns the current time, in seconds, just like ```MPI_Wtime``.  There is a corresponding implementation of ```shmemx_wtime()``` in the SOS implementation that should be updated if we make any changes here (the code you commented on here is not actually exercised when you build this benchmark using SOS).

There is a secondary issue to the loss of precision you noted, which is that ```gettimeofday``` is a poor way to measure the passage of time (see, https://blog.habets.se/2010/09/gettimeofday-should-never-be-used-to-measure-time).  We can use the POSIX timer, ```clock_gettime``` with ```CLOCK_REALTIME``` or ```CLOCK_MONOTONIC``` when it's available and fall back to ```gettimeofday``` otherwise.

```shmemx_time``` was a quick hack from a while back.  Good timers will take some configury and system-specific hacking.  We should be conscious of scope creep..

File: test/performance/shmem_perf_suite/shmem_bibw_nb_get_perf.c (hunk 1 above)
Comment (jdinan) [inline test/performance/shmem_perf_suite/shmem_bibw_nb_get_perf.c]: Should the barrier go before the beginning of each trial?  If we are not doing it already, we could also look at the time from both PEs in the bidirectional test so we are reporting on end-to-end completion.

File: test/performance/shmem_perf_suite/common.h (hunk 3 above)
Comment (jdinan) [inline test/performance/shmem_perf_suite/common.h]: #195

Comment (sayantansur) [inline test/performance/shmem_perf_suite/common.h]: OK sounds good.

File: test/performance/shmem_perf_suite/shmem_bibw_nb_get_perf.c (hunk 1 above)
Comment (sayantansur) [inline test/performance/shmem_perf_suite/shmem_bibw_nb_get_perf.c]: Don't the shmem semantics force you to look at end-to-end completion? Another option might be to do a barrier at the end of the warmups?

We can proceed for now if you're kind enough to create an issue so we don't forget to reinvestigate this on some fast network :-)

Comment (jdinan) [inline test/performance/shmem_perf_suite/shmem_bibw_nb_get_perf.c]: I think we are missing a ```quiet``` here to complete those nonblocking get operations.

The end-to-end completion that I was thinking of is waiting for the gets to complete at both PEs in the bidirectional test.

Comment (sayantansur) [inline test/performance/shmem_perf_suite/shmem_bibw_nb_get_perf.c]: Duh. Yea, you're right.

Comment (kseager) [inline test/performance/shmem_perf_suite/shmem_bibw_nb_get_perf.c]: added in the quiet, thanks for the catch
